Add the ImageClassificationPipeline

LysandreJik opened the pull request:
This PR adds the `ImageClassificationPipeline`. It is tested on DeiT and ViT and should enable the inference API for these models.

Since I encountered an issue with the `AutoExtractor`, I fixed it as seen with @sgugger (namely switched it from using names such as `"deit"` to using the configuration class as a key, similarly to what we do in tokenizers and the other auto classes.

Please let me know if you would like me to split this PR into multiple PRs for simpler reviews, happy to do so.

@NielsRogge, happy to have your review.

LysandreJik commented on src/transformers/feature_extraction_utils.py:
```diff
@@ -226,7 +226,7 @@ def from_pretrained(
                   ``./my_model_directory/``.
                 - a path or url to a saved feature extractor JSON `file`, e.g.,
-                  ``./my_model_directory/feature_extraction_config.json``.
+                  ``./my_model_directory/preprocessor_config.json``.
```
Changed this to the actual name of the configuration file

LysandreJik commented on src/transformers/models/auto/feature_extraction_auto.py:
```diff
@@ -14,34 +14,24 @@
+        (DeiTConfig, DeiTFeatureExtractor),
+        (Speech2TextConfig, Speech2TextFeatureExtractor),
+        (ViTConfig, ViTFeatureExtractor),
+        (Wav2Vec2Config, Wav2Vec2FeatureExtractor),
```
As seen with @sgugger

LysandreJik commented on src/transformers/pipelines/__init__.py:
```diff
@@ -359,19 +383,7 @@ def pipeline(
         # At that point framework might still be undetermined
         model = get_default_model(targeted_task, framework, task_options)
 
-    # Try to infer tokenizer from model or config name (if provided as str)
```
Moved the tokenizer initialization logic beneath the model initialization, with the feature processor. This way it benefits from having access to the model configuration.

LysandreJik commented on src/transformers/pipelines/base.py:
```diff
@@ -522,7 +523,8 @@ class Pipeline(_ScikitCompat):
         self,
         model: Union["PreTrainedModel", "TFPreTrainedModel"],
-        tokenizer: PreTrainedTokenizer,
+        tokenizer: Optional[PreTrainedTokenizer] = None,
```
The `tokenizer` becomes an optional value, alongside the `feature_extractor`.

LysandreJik commented on src/transformers/pipelines/base.py:
```diff
@@ -630,7 +639,14 @@ def check_model_type(self, supported_models: Union[List[str], dict]):
                 The list of models supported by the pipeline, or a dictionary with model class values.
         """
         if not isinstance(supported_models, list):  # Create from a model mapping
-            supported_models = [item[1].__name__ for item in supported_models.items()]
+            supported_models_names = []
+            for config, model in supported_models.items():
+                # Mapping can now contain tuples of models for the same configuration.
+                if isinstance(model, tuple):
+                    supported_models_names.extend([_model.__name__ for _model in model])
+                else:
+                    supported_models_names.append(model.__name__)
+            supported_models = supported_models_names
```
This was forgotten with the addition of the possibility to have two different models for a single configuration in https://github.com/huggingface/transformers/pull/11150.

sgugger replied:
Good catch!

LysandreJik commented on src/transformers/models/auto/feature_extraction_auto.py:
```diff
@@ -14,34 +14,24 @@
+import os
 from collections import OrderedDict
 
+from transformers import DeiTFeatureExtractor, Speech2TextFeatureExtractor, ViTFeatureExtractor
```
This enables the use of dummy objects here rather than having `None` objects

sgugger commented on src/transformers/pipelines/base.py:
```diff
@@ -565,7 +568,13 @@ def save_pretrained(self, save_directory: str):
+
+        if self.feature_extractor is not None:
+            self.feature_extractor.save_pretrained(save_directory)
+
```
Should we have a defensive check when both are None?

sgugger reviewed (approved):
Nice addition! Thanks for working on this!

patil-suraj reviewed (approved):
Nice! Thank you for adding this, LGTM.

patrickvonplaten commented on src/transformers/models/auto/feature_extraction_auto.py:
```diff
@@ -134,20 +124,29 @@ def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
+
+        is_feature_extraction_file = os.path.isfile(pretrained_model_name_or_path)
+        is_directory = os.path.isdir(pretrained_model_name_or_path) and os.path.exists(
+            os.path.join(pretrained_model_name_or_path, "preprocessor_config.json")
```
Can we maybe instead use the constant `FEATURE_EXTRACTOR_NAME` from `src/transformes/file_utils.py` -> don't think we will change this file name again, but seems safer.

patrickvonplaten commented on src/transformers/models/auto/feature_extraction_auto.py:
```diff
@@ -134,20 +124,29 @@ def from_pretrained(cls, pretrained_model_name_or_path, **kwargs):
+        elif "feature_extractor_type" in config_dict:
             feature_extractor_class = feature_extractor_class_from_name(config_dict["feature_extractor_type"])
             return feature_extractor_class.from_dict(config_dict, **kwargs)
-        else:
```
nice!

patrickvonplaten commented on src/transformers/pipelines/__init__.py:
```diff
@@ -309,6 +321,18 @@ def pipeline(
+            identifier or an actual pretrained feature extractor inheriting from
+            :class:`~transformers.PreTrainedFeatureExtractor`.
+
+            Feature extractors are used for non-NLP models, such as Speech or Vision models. If multi-modal, will also
```
(nit)
```suggestion
            Feature extractors are used for non-NLP models, such as Speech or Vision models as well as multi-modal models.
```

patrickvonplaten commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+@add_end_docstrings(PIPELINE_INIT_ARGS)
+class ImageClassificationPipeline(Pipeline):
+    """
+    Image classification pipeline using any :obj:`ModelForImageClassification`. This pipeline predicts the class of an
```
(nit)
```suggestion
    Image classification pipeline using any :obj:`AutoModelForImageClassification`. This pipeline predicts the class of an
```

patrickvonplaten commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+                The pipeline accepts either a single image or a batch of images, which must then be passed as a string.
+                Images in a batch must all be in the same format: all as http links, all as local paths, or all as PIL
+                images.
+            top_k (:obj:`int`):
```
```suggestion
            top_k (:obj:`int`, `optional`, defaults to5):
```

patrickvonplaten commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+            - **label** (:obj:`str`) -- The label identified by the model.
+            - **score** (:obj:`int`) -- The score attributed by the model for that label.
+        """
+        is_batched = type(images) == list
```
just out of curiosity - is it not possible/supported to have multiple images saved in the PIL format? It is necessarily `List[PIL.image]` and not just `PIL.image`?

NielsRogge replied:
Yes they have to be a `List[PIL.Image]`. Didn't know you can save multiple images in a single `PIL.Image`.

patrickvonplaten commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+            images = [Image.open(image) for image in images]
+
+        inputs = self.feature_extractor(images=images, return_tensors="pt")
+        outputs = self.model(**inputs)
```
wrap in `torch.no_grad():` no? Otherwise activations are stored which would make the forward pass much more expensive I think

patrickvonplaten commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+        probs = outputs.logits.softmax(-1)
+        scores, ids = probs.topk(top_k)
+
+        scores = scores.detach().tolist()
```
```suggestion
        scores = scores.tolist()
```

patrickvonplaten commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+        scores, ids = probs.topk(top_k)
+
+        scores = scores.detach().tolist()
+        ids = ids.detach().tolist()
```
```suggestion
        ids = ids.tolist()
```
shouldn't be needed if wrapped into `torch.no_grad():`

patrickvonplaten commented on tests/test_feature_extraction_auto.py:
```diff
@@ -29,16 +30,10 @@ def test_feature_extractor_from_model_shortcut(self):
         config = AutoFeatureExtractor.from_pretrained(SAMPLE_FEATURE_EXTRACTION_CONFIG)
         self.assertIsInstance(config, Wav2Vec2FeatureExtractor)
-
-    def test_pattern_matching_fallback(self):
```
nice!

patrickvonplaten commented on tests/test_pipelines_image_classification.py:
```diff
@@ -0,0 +1,77 @@
+    class Image:
+        @staticmethod
+        def open(*args, **kwargs):
+            pass
```
I wonder whether it might be better to provide a warning here so that one knows it's a dummy class and wonders what happened to the `PIL.Image.open()`

LysandreJik replied:
This is a dummy class that's only useful for testing purposes (it's in a test class). The class property `valid_input` is called when the class is created, therefore running the following code: `Image.open("tests/fixtures/coco.jpg")`. It does so even when PIL isn't installed, which would then crash were a suite run without `PIL` installed.

patrickvonplaten reviewed (approved):
Nice!

The only thing, I feel quite strongly about is to wrap the forward pass into a `with torch.no_grad():` to avoid computing costly activations. 

The rest is all nits

NielsRogge commented:
LGTM, thank you for adding this.

Narsil commented on tests/test_pipelines_image_classification.py:
```diff
@@ -0,0 +1,77 @@
+    ]
+
+    def setUp(self) -> None:
+        if not is_torch_available():
```
Also we can keep the actual many calls in  `small_model` and limit the amount of classification made with large models (which should be there only to ensure we have correct results and that they don't change over time)

LysandreJik replied:
I agree with you, will update the code accordingly. Thanks!

Narsil commented on src/transformers/pipelines/__init__.py:
```diff
@@ -434,6 +433,65 @@ def pipeline(
+                tokenizer_identifier, revision=revision, use_fast=use_fast, _from_pipeline=task, **tokenizer_kwargs
+            )
+    else:
+        tokenizer = None
```
That whole logic seems extremely convoluted and error prone.

- You override the `tokenizer` to None if `load_tokenizer == False`. That's a very bad behavior for this function in my opinion. If I pass you an already instantiated class, you should just pass it as it, you should never override it (Imagine someone implemented their own tokenizer)
- `tokenizer_kwargs = model_kwargs` seems super risky. They are not the same, this is bound to cause some errors at some point.
- We could probably make the intent easier to understand too by using different names. for instance like `model_name` is necessarily a string we could have `tokenizer_name` (or `tokenizer_identifier`) 
- `tokenizer` argument should never be a tuple (it's not stated in the function signature at least, and it seems error prone) (This was already there, so we should keep it, but still a wart no ?)

LysandreJik replied:
You're right about overriding the tokenizer to `None` - the rest were already there in some shape or form but happy to cleanup in this PR. Will propose changes.

Narsil replied:
Yes I know some stuff was already there. The biggest point as you noted is silent override of the argument.

This function is a challenge to modify ;)

LysandreJik replied:
In the end I modified the `tokenizer` set to `None`, and left the rest as it initially was. As said above, happy to tackle this in a cleanup PR.

I added a test to ensure that we could pass a custom tokenizer per the use-case you've mentioned.

Narsil reviewed (changes requested):
Biggest change is the `tokenizer` logic modification, we should really respect that if a user sends us something that is not a `str`(or `tuple`...) then we need to assume it's a valid tokenizer and pass it as is.

Also the tests are a bit bloated IMO compared to what they could be. In that instance, because we can't test parity with TF, we should just make the tests much simpler.
Also we don't test pipeline outputs which is a big issue for maintaining BC.

The image loading logic should be fixed too.

The rest are NITs.

Narsil commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+            images = [images]
+
+        if type(images[0]) == str and images[0].startswith("http"):
+            images = [Image.open(requests.get(image, stream=True).raw) for image in images]
```
Edit, just thought, we should probably try `isfile` before checking for `https://` because IIRC `/` is valid on windows filenames.

LysandreJik replied:
> This is a bad mistake, you are only checking for the first argument's type, and treating all the others as if they were the same.

I wouldn't say it's a "bad mistake", but rather a design choice :) I have put the following in the docs: `Images in a batch must all be in the same format: all as http links, all as local paths, or all as PIL images.`

I understand how you think this is a bad design decision however, happy to update to the code you propose.

---

Regarding the test with `isfile` before raising a custom `ValueError`, I think the error raised by `PIL` is actually very explicit
```
FileNotFoundError: [Errno 2] No such file or directory: 'xxx'
```

What would you rather have?

Narsil replied:
Ok letting PIL handle the error is OK not. My main concern was "http" detection and doing sanitation on all inputs.

LysandreJik commented on src/transformers/pipelines/__init__.py:
```diff
@@ -434,6 +433,65 @@ def pipeline(
             model, config=config, revision=revision, _from_pipeline=task, **model_kwargs
         )
 
+    model_config = model.config
```
Absolutely, `config` is an optional argument which could be `None`. Taking the `model.config` ensures we have the actual model configuration.

LysandreJik commented on src/transformers/pipelines/base.py:
```diff
@@ -522,7 +523,8 @@ class Pipeline(_ScikitCompat):
         model: Union["PreTrainedModel", "TFPreTrainedModel"],
-        tokenizer: PreTrainedTokenizer,
+        tokenizer: Optional[PreTrainedTokenizer] = None,
+        feature_extractor: Optional[PreTrainedFeatureExtractor] = None,
```
Good point! How should `tokenizer` be handled then? `tokenizer` will only be handled in text-only pipelines, and not in some multi-modal pipelines such as the image classification pipeline. Would you like to remove this argument here?

Narsil replied:
Well the only drawback for  `tokenizer` is backward compatibility.

There's probably not a lot of users of the raw `Pipeline` but still. We should check that all subclasses actually declare __init__ correctly too. Maybe better left for a cleanup PR ?

LysandreJik replied:
Happy to explore that in a cleanup PR.

LysandreJik commented on tests/test_pipelines_image_classification.py:
```diff
@@ -0,0 +1,77 @@
+                )
+
+    def _test_pipeline(self, image_classifier: Pipeline):
+        image_classifier(
```
Good point, will update. Thanks!

LysandreJik commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+
+        self.feature_extractor = feature_extractor
+
+    def save_pretrained(self, save_directory: str):
```
Agree!

LysandreJik commented on src/transformers/pipelines/image_classification.py:
```diff
@@ -0,0 +1,132 @@
+        framework: Optional[str] = None,
+        **kwargs
+    ):
+        super().__init__(model, framework=framework, **kwargs)
```
This is my mistake indeed, started off by not having `feature_processor` be part of the base class, hence the swallowing.

LysandreJik commented:
I should have addressed all comments. 

@Narsil, if you could review once again when you have time, would love your feedback on the updated tests and on the updated `__init__` method.

Narsil reviewed (approved):
LGTM.

Last nit, maybe the `load_image`  function could be put not inlined (like at the module level).

It should make it easier to test and overload.
